src/main/java/software/amazon/awssdk/crt/CrtResource.java

```diff
@@ -175,11 +177,21 @@ protected void swapReferenceTo(CrtResource oldReference, CrtResource newReferenc
         }
     }
 
+    protected void acquireReadLock() {
+        this.rwl.readLock().lock();
+    }
+
+    protected void releaseReadLock() {
+        this.rwl.readLock().unlock();
+    }
+
+
     /**
      * Takes ownership of a native object where the native pointer is tracked as a long.
      * @param handle pointer to the native object being acquired
      */
     protected void acquireNativeHandle(long handle) {
+        this.rwl.writeLock().lock();
```

Comment:
Use some mechanism to ensure a random exception doesn't bounce us out of this block before unlock

try with resources
try/finally
something

Reply:
The advantage of `synchronized` is: it's dead simple. The only way to mess up is to forget to use it.

If we used `synchronized` we could add CI to catch where we forget to use it. We could like use Java reflection to scan all subclasses of CrtResource, scan all methods, check if [method.isSynchronized()](https://docs.oracle.com/javase%2F9%2Fdocs%2Fapi%2F%2F/com/sun/jdi/Method.html#isSynchronized--), and error if so (maybe there's be an allow-list for functions that didn't need to be synchronized?